llama.cpp

```diff
@@ -1416,26 +1423,39 @@ static bool llama_eval_internal(
     auto & mem_per_token = lctx.mem_per_token;
     auto & buf_compute   = lctx.buf_compute;
 
+
     struct ggml_init_params params = {
         /*.mem_size   =*/ buf_compute.size,
         /*.mem_buffer =*/ buf_compute.addr,
         /*.no_alloc   =*/ false,
     };
 
+#ifdef LLAMA_USE_ALLOCATOR
+#  define ggml_rope_custom_inplace ggml_rope_custom
+#  define ggml_scale_inplace ggml_scale
+#  define ggml_diag_mask_inf_inplace ggml_diag_mask_inf
+#  define ggml_soft_max_inplace ggml_soft_max
+    params.no_alloc = true;
+#endif
```

Comment:
What happens if the ops are `inplace`?

Reply:
The problem with `inplace` ops is that the allocator needs to know what tensors are views of another tensor to be able to know when it is safe to free a tensor - ie. a tensor cannot be freed as along there are views of it. `inplace` ops are effectively a view of their parent, and the allocator is not able to detect that.

In this case however, it wouldn't really matter since the context is `no_alloc`, the parent's `data` will be `NULL`, so the "inplace" op will also have a `NULL` pointer, and the result would be indistinguishable from a normal op. So these defines could be removed. The allocator could also try harder to determine if an op is `inplace` by checking if its `data` is the same as one of its parents. The code could also be simplified a bit if we simply added a pointer to the source of the view in `ggml_tensor`, though.

The allocator will already make all suitable ops `inplace` automatically to save memory.